client/flaschen.py

```diff
@@ -0,0 +1,66 @@
+# -*- mode: python; c-basic-offset: 2; indent-tabs-mode: nil; -*-
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation version 2.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+import socket
+
+class Flaschen(object):
+  '''A Framebuffer display interface that sends a frame via UDP.'''
+
+  def __init__(self, host, port, width, height, layer=0, transparent=False):
+    '''
+
+    Args:
+      host: The flaschen taschen server hostname or ip address.
+      port: The flaschen taschen server port number.
+      width: The width of the flaschen taschen display in pixels.
+      height: The height of the flaschen taschen display in pixels.
+      layer: The layer of the flaschen taschen display to write to.
+      transparent: If true, black(0, 0, 0) will be transparent and show the layer below.
+    '''
+    self.width = width
+    self.height = height
+    self.layer = layer
+    self.transparent = transparent
+    self.pixels = bytearray(width * height * 3)
+    self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    self.sock.connect((host, port))
+    self._header = ''.join(["P6\n",
+                            "%d %d\n" % (self.width, self.height),
+                            "255\n"])
+    self._footer = ''.join(["0\n",
+                            "0\n",
+                            "%d\n" % self.layer])
+
+  def set(self, x, y, color):
+    '''Set the pixel at the given coordinates to the specified color.
+
+    Args:
+      x: x offset of the pixel to set
+      y: y offset of the piyel to set
+      color: A 3 tuple of (r, g, b) color values, 0-255
+    '''
+    if x >= self.width or y >= self.height or x < 0 or y < 0:
+      return
+    if color == (0, 0, 0) and not self.transparent:
+      color = (1, 1, 1)
+
+    offset = (x + y * self.width) * 3
+    self.pixels[offset] = color[0]
+    self.pixels[offset + 1] = color[1]
+    self.pixels[offset + 2] = color[2]
+  
+  def send(self):
+    '''Send the updated pixels to the display.'''
+    data = self._header + self.pixels + "\n" + self._footer
```

Comment:
So, here you are doing an expensive concatenation operation where we need to do allocation, garbage collection etc.

How about allocating that in the constructor. Creating a dummy byte-array there of the right pixel size, and do exactly the concatenation you do here in `send()` once in the constructor and remember the result. All you remember is the resulting data[] array and the length of the header (_header and _footer itself don't even have to be member variables at that point anymore as you don't need them unless providing a layer-setting somewhere).

In the set() method, you just do self._data[self._header_offset + pixel_offset], in the `send()` method you just send the buffer: self.socket.send(self._data)